source/geod24/concurrency.d

```diff
@@ -985,6 +986,161 @@ protected:
         bool notified;
     }
 
+    class Channel (T)
+    {
+        this(ulong size = 0)
+        {
+            this.size = size;
+            m_lock = new Mutex;
+        }
+
+        bool write(T val)
+        {
+            bool notify_peer, fiber_was_blocked = false;
+            ChannelQueueEntry qEnt;
+
+            synchronized (m_lock)
+            {
+                notify_peer = m_rxq.length > 0;
+
+                if (m_closed)
+                {
+                    return false;
+                }
+                else if (m_rxq.length > 0)
+                {
+                    qEnt = m_rxq[0];
+                    m_rxq = m_rxq.remove(0);
+
+                    *(qEnt.pVal) = val;
+                }
+                else if (m_buffer.length < size)
+                {
+                    m_buffer ~= val;
+                }
+                else
+                {
+                    qEnt = new ChannelQueueEntry(&val);
+                    m_txq ~= qEnt;
+                }
+            }
+
+            if (notify_peer)
+            {
+                qEnt.sem.notify();
+            }
+            else if (qEnt)
+            {
+                fiber_was_blocked = true;
+                qEnt.sem.wait();
+            }
+
+            return !(fiber_was_blocked) || qEnt.success;
+        }
+
+        bool read(ref T output)
+        {
+            bool notify_peer, fiber_was_blocked = false;
+            ChannelQueueEntry qEnt;
+
+            synchronized (m_lock)
+            {
+                notify_peer = m_txq.length > 0;
+
+                if (m_buffer.length != 0)
+                {
+                    output = m_buffer[0];
+                    m_buffer = m_buffer.remove(0);
+
+                    if (m_txq.length > 0)
+                    {
+                        qEnt = m_txq[0];
+                        m_txq = m_txq.remove(0);
+
+                        m_buffer ~= *(qEnt.pVal);
+                    }
+                }
+                else if (m_closed)
```

Comment:
How come this isn't the first check that we do, but only the second one? The buffer must be flushed?

Reply:
Current implementation lets already buffered messages to be read after the channel is closed. Do we have any preference in that?

Reply:
Hmm, I think we don't have a preference. What does golang do?

Reply:
Just tested it, Golang lets existing messages to be read after the channel gets closed.

Reply:
I have a preference towards letting messages be read. So you can send an "end notification" and close immediately, instead of having to send an "end notification" and wait for an ACK before closing the channel (or have the other party close it).